File: Problem 4/largestPalindrome.py

def findLargestPalindrome(numA, numB, largestPalProduct):
    # Keep testing products until Number A and Number B are no longer 3 digits
    # long
    while len(str(numA)) < 4:
        while len(str(numB)) < 4:
            # Multiplying the Number A and Number B together, to find the
            # product
            palTest = numA * numB

            # Splitting the string into two by incrementing over it using a
            # pair of slicers. This gets us numPartA and numPartB
            numPartA, numPartB = str(palTest)[:int(len(str(palTest))/2)], str(palTest)[int(len(str(palTest))/2):]

            # We reverse Part B
            numPartB = numPartB[::-1]

            # Then we test to see if the two parts are the same. If they are,
            # then we check to see if the new palindrome is larger than the
            # currently stored largest. If it is, we replace it, if not, we
            # just continue onto the next loop
            if numPartA == numPartB:
                if palTest > largestPalProduct:
                    largestPalProduct = palTest

            # Incrementing Number B
            numB += 1

        # Incrementing Number A and resetting Number B
        numA += 1
        numB = 100

    return largestPalProduct


def main():
    numA = 100
    numB = 100

    print(findLargestPalindrome(numA, numB, 0))


main()

File: Problem 4/largestPalindrome.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;


void findLargestPal (int numA, int numB, double largestPal) {

  //cout << "Find Largest Pal  RUnning\n";

  cout << "Number A BEFORE: " + to_string(numA) + "\n";
  cout << "Number B BEFORE " + to_string(numB) + "\n";

  // We keeping testing NumA and NumB until they are longer than 3 digits
  while (numA <= 999) {
    while (numB <= 999) {
      // Multipling the NumA and B together to find their product
      int palTest = numA * numB;

      string tempStr = to_string(palTest);

      cout << "\n"
           << "Sum Result: " + to_string(palTest)
           << "\n";

      // Converting the palTest result into a string, and using substr to
      // split the string into two halves. The string will ALWAYS be 6 digits,
      // So we can hard code the numbers in
      string numPartA = to_string(palTest).substr(0, 3);
      cout << "Pal Test After Subtraction: " + to_string(palTest) + "\n";
      string numPartB = to_string(palTest).substr(4, 6);

      cout << "Number Part A: " + numPartA + "\n";
      cout << "Number Part B: " + numPartB + "\n";

      // Reversing the second  half of the split to be compared
      reverse(begin(numPartB), end(numPartB));

      // Comparing the two halves together, to see if they're the same
      if (numPartA == numPartB) {
        // If they are the same, we then see if the sum of the two parts is
        // larger than the currently largest sum, we replace it
        if (palTest > largestPal) {
          cout << "Currently Largest Pal: " + to_string(largestPal) + "\n";
          cout << "Testing Pal: " + to_string(palTest) + "\n";
          largestPal = palTest;
        }
        // Increment NumB
        numB++;
      }
      // Incrementing numA and Resetting numB
      numA++;
      numB = 100;
    }
    // Return the result
    cout << to_string(largestPal) + "\n";
  }
}


int main () {
  int numA = 100;
  int numB = 100;

  findLargestPal(numA, numB, 0);

  cout << "Result!!";


}
